retroshare-gui/src/gui/common/FriendListModel.cpp

```diff
@@ -0,0 +1,1101 @@
+
+            return QVariant(last_contact);
+		}
+            break;
+        default:
+            return QVariant();
+		}
+    }
+        break;
+
+    default:
+		return displayRole(entry,column);
+    }
+}
+
+QVariant RsFriendListModel::onlineRole(const EntryIndex& e, int col) const
+{
+    switch(e.type)
+	{
+    default:
+	case ENTRY_TYPE_GROUP:
+    {
+        const HierarchicalGroupInformation& g(mGroups[e.group_index]);
+
+        for(uint32_t j=0;j<g.child_profile_indices.size();++j)
+		{
+			const HierarchicalProfileInformation& prof = mProfiles[g.child_profile_indices[j]];
+
+			for(uint32_t i=0;i<prof.child_node_indices.size();++i)
+				if(mLocations[prof.child_node_indices[i]].node_info.state & RS_PEER_STATE_CONNECTED)
+					return QVariant(true);
+		}
+        return QVariant(false);
+    }
+
+	case ENTRY_TYPE_PROFILE:
+    {
+		const HierarchicalProfileInformation *prof = getProfileInfo(e);
+
+        if(!prof)
+            return QVariant();
+
+        for(uint32_t i=0;i<prof->child_node_indices.size();++i)
+            if(mLocations[prof->child_node_indices[i]].node_info.state & RS_PEER_STATE_CONNECTED)
+                return QVariant(true);
+
+        return QVariant();
+    }
+        break;
+
+	case ENTRY_TYPE_NODE:
+        const HierarchicalNodeInformation *node = getNodeInfo(e);
+
+        if(!node)
+            return QVariant();
+
+        return QVariant(bool(node->node_info.state & RS_PEER_STATE_CONNECTED));
+	}
+}
+
+QVariant RsFriendListModel::fontRole(const EntryIndex& e, int col) const
+{
+#ifdef DEBUG_MODEL
+	std::cerr << "  font role " << e.type << ", (" << (int)e.group_index << ","<< (int)e.profile_index << ","<< (int)e.node_index << ") col="<< col<<": " << std::endl;
+#endif
+
+    bool b = onlineRole(e,col).toBool();
+
+    if(b)
+    {
+        QFont font ;
+		font.setBold(b);
+
+        return QVariant(font);
+    }
+    else
+        return QVariant();
+}
+
+class AutoEndel
+{
+public:
+    ~AutoEndel() { std::cerr << std::endl;}
+};
+
+QVariant RsFriendListModel::displayRole(const EntryIndex& e, int col) const
+{
+#ifdef DEBUG_MODEL
+    std::cerr << "  Display role " << e.type << ", (" << (int)e.group_index << ","<< (int)e.profile_index << ","<< (int)e.node_index << ") col="<< col<<": ";
+    AutoEndel x;
+#endif
+
+    switch(e.type)
+	{
+	case ENTRY_TYPE_GROUP:
+        {
+  	      const HierarchicalGroupInformation *group = getGroupInfo(e);
+
+          uint32_t nb_online = 0;
+
+          for(uint32_t i=0;i<group->child_profile_indices.size();++i)
+              for(uint32_t j=0;j<mProfiles[group->child_profile_indices[i]].child_node_indices.size();++j)
+                  if(mLocations[mProfiles[group->child_profile_indices[i]].child_node_indices[j]].node_info.state & RS_PEER_STATE_CONNECTED)
+                  {
+                      nb_online++;
+                      break;// only breaks the inner loop, on purpose.
+                  }
+
+  	      if(!group)
```

Comment:
Shouldn't this check go right after `getGroupInfo(e);` if `grpup` is `null` you will likely crash in the loop otherwise.

Reply:
right!